google/cloud/storage/object_access_control.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/object_access_control.h"
+#include "google/cloud/storage/internal/metadata_parser.h"
+#include "google/cloud/storage/internal/nljson.h"
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+ObjectAccessControl ObjectAccessControl::ParseFromJson(
+    std::string const& payload) {
+  auto json = storage::internal::nl::json::parse(payload);
+  ObjectAccessControl result{};
+  result.bucket_ = json.value("bucket", "");
+  result.domain_ = json.value("domain", "");
+  result.email_ = json.value("email", "");
+  result.entity_ = json.value("entity", "");
+  result.entity_id_ = json.value("entityId", "");
+  result.etag_ = json.value("etag", "");
+  result.generation_ =
+      internal::MetadataParser::ParseLongField(json, "generation");
+  result.id_ = json.value("id", "");
+  result.kind_ = json.value("kind", "");
+  result.object_ = json.value("object", "");
+  result.role_ = json.value("role", "");
+  result.self_link_ = json.value("selfLink", "");
+  if (json.count("projectTeam") != 0U) {
+    auto tmp = json["projectTeam"];
+    result.project_team_.project_number = tmp.value("projectNumber", "");
+    result.project_team_.team = tmp.value("team", "");
+  }
+  return result;
+}
+
+bool ObjectAccessControl::operator==(ObjectAccessControl const& rhs) const {
+  // Start with id, generation, object, bucket, etag because they should fail
```

Comment:
I'm not sure I follow what you mean about failing early. Are you just saying they're the most likely culprits for = mismatch, so checking them first is more efficient?

Reply:
Yes.  I do not have strong opinions about this, but @houglum felt it was important in a similar class and I am following the same pattern.